File: rmw_launch_smoke_test/subscriber.py
#!/usr/bin/env python3
# Copyright 2026 Daisuke Kato
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import time

import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node

from std_msgs.msg import String


class SmokeSubscriber(Node):
    """Receive five ordered messages or fail after a bounded timeout."""

    def __init__(self):
        super().__init__('smoke_subscriber')
        self._received = 0
        self._next_message = 0
        self._failed = False
        self._deadline = time.monotonic() + 20.0
        self._subscription = self.create_subscription(
            String, 'smoke_chatter', self._receive, 10)
        self._watchdog = self.create_timer(0.1, self._check_progress)

    def _receive(self, message):
        expected = f'hello-{self._next_message}'
        if message.data != expected:
            self.get_logger().error(
                f'expected={expected}, received={message.data}')
            self._failed = True
            rclpy.shutdown()
            return

        self._received += 1
        self._next_message += 1
        self.get_logger().info(f'received={self._received}')
        if self._received == 5:
            self._watchdog.cancel()
            rclpy.shutdown()

    def _check_progress(self):
        if time.monotonic() > self._deadline:
            self.get_logger().error(
                f'timed out after receiving {self._received} messages')
            self._failed = True
            self._watchdog.cancel()
            rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    node = SmokeSubscriber()
    try:
        rclpy.spin(node)
    except ExternalShutdownException:
        pass
    finally:
        failed = node._failed
        node.destroy_node()
    return 1 if failed else 0


if __name__ == '__main__':
    raise SystemExit(main())

File: rmw_launch_smoke_test_cpp/src/subscriber.cpp
// Copyright 2026 Daisuke Kato
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class SmokeSubscriber : public rclcpp::Node
{
public:
  SmokeSubscriber()
  : Node("smoke_subscriber"), received_(0), next_message_(0), failed_(false),
    deadline_(std::chrono::steady_clock::now() + 20s)
  {
    subscription_ = create_subscription<std_msgs::msg::String>(
      "smoke_chatter", 10,
      [this](std_msgs::msg::String::ConstSharedPtr message) {receive(message);});
    watchdog_ = create_wall_timer(100ms, [this]() {check_progress();});
  }

  bool failed() const {return failed_;}

private:
  void receive(std_msgs::msg::String::ConstSharedPtr message)
  {
    const auto expected = "hello-" + std::to_string(next_message_);
    if (message->data != expected) {
      RCLCPP_ERROR(
        get_logger(), "expected=%s, received=%s", expected.c_str(), message->data.c_str());
      failed_ = true;
      rclcpp::shutdown();
      return;
    }

    ++received_;
    ++next_message_;
    RCLCPP_INFO(get_logger(), "received=%d", received_);
    if (received_ == 5) {
      watchdog_->cancel();
      rclcpp::shutdown();
    }
  }

  void check_progress()
  {
    if (std::chrono::steady_clock::now() > deadline_) {
      RCLCPP_ERROR(get_logger(), "timed out after receiving %d messages", received_);
      failed_ = true;
      watchdog_->cancel();
      rclcpp::shutdown();
    }
  }

  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
  rclcpp::TimerBase::SharedPtr watchdog_;
  int received_;
  int next_message_;
  bool failed_;
  std::chrono::steady_clock::time_point deadline_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SmokeSubscriber>();
  rclcpp::spin(node);
  const bool failed = node->failed();
  rclcpp::shutdown();
  return failed ? 1 : 0;
}
